src/libos/crates/async-rt/src/executor.rs

```diff
@@ -170,3 +377,302 @@ impl Executor {
         self.is_shutdown.load(Ordering::Relaxed)
     }
 }
+
+pub(crate) struct Worker {
+    high_pri_queue: TaskQueueWithSlot,
+    normal_pri_queue: TaskQueueWithSlot,
+    low_pri_queue: TaskQueueWithSlot,
+}
+
+impl Worker {
+    pub const QUEUED_TASKS_MAX_SIZE: usize = 1_000;
+
+    pub fn new() -> Self {
+        Self {
+            high_pri_queue: TaskQueueWithSlot::new(Some(Self::QUEUED_TASKS_MAX_SIZE)),
+            normal_pri_queue: TaskQueueWithSlot::new(Some(Self::QUEUED_TASKS_MAX_SIZE)),
+            low_pri_queue: TaskQueueWithSlot::new(Some(Self::QUEUED_TASKS_MAX_SIZE)),
+        }
+    }
+
+    pub fn push(&self, task: Arc<Task>, injector: &Injector) -> bool {
+        match task.sched_info().priority() {
+            SchedPriority::High => {
+                if let Err(t) = self.high_pri_queue.push(task) {
+                    injector.push(t);
+                    return false;
+                }
+            }
+            SchedPriority::Normal => {
+                if let Err(t) = self.normal_pri_queue.push(task) {
+                    injector.push(t);
+                    return false;
+                }
+            }
+            SchedPriority::Low => {
+                if let Err(t) = self.low_pri_queue.push(task) {
+                    injector.push(t);
+                    return false;
+                }
+            }
+        }
+        true
+    }
```

Comment:
There are too many duplicate code to handle the three priorities. I think it is good idea to introduce some macros to simplify the code.

```rust
// Apply a block of code on the queue for the given priority
macro_rules! apply_on_queue {
    ($priority:expr, |queue| { $($code:tt)* }) => {
        // implement the macro
    }
}
```

With this macro, the `Worker` code can be greatly simplified.
```rust
impl Worker {
    pub fn push(&self, task: Arc<Task>, injector: &Injector) -> bool {
        apply_on_queue(task.sched_info().priority(), |queue| {
            if let Err(t) = queue.push(task) {
                return Some(task);
            }
        })
    }

    pub fn pop_with_priority(&self, priority: SchedPriority) -> Option<Arc<Task>> {
    	apply_on_queue(priority, |queue| {
            queue.pop()
    	})
    }    

    pub fn checked_pop_with_priority<F>(
        &self,
        f: F,
        priority: SchedPriority,
    ) -> Option<(usize, Arc<Task>)>
    where
        F: FnOnce(&Arc<Task>) -> Option<usize>,
    {
    	apply_on_queue(priority, |queue| {
            queue.checked_pop(f)
    	})
    }

    pub fn len(&self, priority: SchedPriority) -> usize {
    	apply_on_queue(priority, |queue| {
           queue.len()
    	})
    }

    pub fn relax_len(&self, priority: SchedPriority) -> usize {
    	apply_on_queue(priority, |queue| {
            queue.relax_len()
    	})
    }

    pub fn is_empty(&self, priority: SchedPriority) -> bool {
    	apply_on_queue(priority, |queue| {
           queue.is_empty()
    	})
    }  
}
```

Reply:
Second thought. Using macros is an overkill. Simply add function.

```rust
fn queue(priority: SchedPriority) -> &TaskQueueWithSlot { /* match priority to queue */ }
```